File: lib/arquivador/caminhos.ts
import "server-only";
import { normalizarSegmento } from "../documentos/nomenclatura";
import { sanitizarNomeArquivo } from "../documentos/inspecao";

/**
 * Expansão de uma REGRA de arquivamento em caminho e nome concretos.
 *
 * Este módulo é PURO: não fala com o banco nem com o Drive. Só transforma
 * (regra + contexto) em (segmentos, chaves lógicas, nome do arquivo). É onde
 * mora a decisão de "onde isso vai e como se chama" — e por ser puro, é a
 * parte que dá para testar exaustivamente sem rede.
 *
 * Nenhum caminho é hardcoded: tudo vem de `regras_arquivamento.caminho_modelo`
 * e `padrao_nome`, carregados do banco.
 *
 * PLACEHOLDERS DE CAMINHO: {ANO} {COMPETENCIA} {PROJETO}
 * PLACEHOLDERS DE NOME:    {CODIGO} {EMPRESA} {COMPETENCIA} {TIPO_DOCUMENTO}
 *                          {INSTITUICAO} {PROJETO} {DATA_DOCUMENTO} {VERSAO}
 *                          {EXTENSAO}
 *
 * {CODIGO}  = código curto da empresa (ex.: TL)
 * {EMPRESA} = nome normalizado da empresa; cai para o código se não houver.
 *
 * Placeholder opcional que não tem valor é REMOVIDO, e os separadores que
 * sobram são colapsados — nunca sai um `{INSTITUICAO}` literal nem um `__`
 * no meio do nome.
 */

export type EscopoRegra = "interno" | "fixo" | "mensal" | "projeto";

/** Escopo de uma pasta, espelhando o CHECK de pastas_drive.escopo. */
export type EscopoPasta =
  | "raiz"
  | "estrutural"
  | "empresa"
  | "periodo"
  | "categoria"
  | "projeto";

export type RegraArquivamento = {
  id: string;
  codigo: string;
  nome: string;
  escopo: EscopoRegra;
  caminho_modelo: string[];
  padrao_nome: string;
  exige_empresa: boolean;
  exige_competencia: boolean;
  exige_instituicao: boolean;
  projeto: string | null;
  subcategoria: string | null;
};

export type ContextoArquivamento = {
  empresaId?: string | null;
  empresaCodigo?: string | null;
  empresaNome?: string | null;
  /** Nome exato da pasta já existente, resolvido por código da empresa. */
  pastaEmpresa?: string | null;
  /**
   * Nome do contêiner onde a pasta da empresa vive
   * (01_CLIENTES_ATIVOS ou 02_CLIENTES_INATIVOS).
   *
   * Vem de `estrutura_fixa_drive`, escolhido pelo chamador conforme
   * `empresas.ativo`. Não tem valor padrão de propósito: chutar "ativos"
   * arquivaria documento de cliente inativo no lugar errado, em silêncio.
   */
  pastaClientes?: string | null;
  /** Competência no formato AAAA-MM. */
  competencia?: string | null;
  instituicao?: string | null;
  tipoDocumento?: string | null;
  /** Data do documento no formato AAAA-MM-DD. */
  dataDocumento?: string | null;
  versao?: number | null;
  extensao?: string | null;
  projeto?: string | null;
};

/** Um nível da árvore de pastas, já com a identidade que vai para o banco. */
export type SegmentoDestino = {
  nome: string;
  escopo: EscopoPasta;
  /** Identidade estável da pasta. Renomear a empresa não quebra o vínculo. */
  chaveLogica: string;
  /** Caminho acumulado até este nível, para leitura humana. */
  caminhoLogico: string;
};

export type DestinoResolvido = {
  ok: true;
  segmentos: SegmentoDestino[];
  /** Caminho completo da pasta final, relativo à raiz do Drive. */
  caminhoLogico: string;
  /** Chave lógica da pasta final — a que o resolvedor usa. */
  chaveLogica: string;
};

export type FalhaDestino = {
  ok: false;
  erro: string;
  /** Campos do contexto que faltaram, para o agente saber o que perguntar. */
  faltando: string[];
};

const FORMATO_COMPETENCIA = /^\d{4}-(0[1-9]|1[0-2])$/;
const FORMATO_DATA = /^\d{4}-(0[1-9]|1[0-2])-(0[1-9]|[12]\d|3[01])$/;

/** Nome da pasta da empresa. Prefere o código curto; cai para o nome. */
function nomePastaEmpresa(ctx: ContextoArquivamento): string {
  if (ctx.pastaEmpresa) return sanitizarNomeArquivo(ctx.pastaEmpresa);
  return normalizarSegmento(ctx.empresaCodigo ?? ctx.empresaNome ?? "");
}

/**
 * Valida o contexto contra o que a regra exige.
 *
 * Devolve a LISTA do que falta em vez de lançar: o agente precisa saber
 * exatamente o que perguntar ao usuário, não só que deu errado.
 */
function conferirExigencias(
  regra: RegraArquivamento,
  ctx: ContextoArquivamento
): string[] {
  const faltando: string[] = [];

  if (regra.exige_empresa) {
    if (!ctx.empresaId) faltando.push("empresaId");
    if (!ctx.empresaCodigo && !ctx.empresaNome) faltando.push("empresaCodigo");
    // Sem o contêiner não dá para montar o caminho: a pasta da empresa cairia
    // solta na raiz do Drive, ao lado de 00_INTERNO.
    if (!ctx.pastaClientes) faltando.push("pastaClientes");
  }

  if (regra.exige_competencia && !ctx.competencia) faltando.push("competencia");
  if (regra.exige_instituicao && !ctx.instituicao) faltando.push("instituicao");

  // Só exigimos projeto se o modelo realmente usar o placeholder — as regras
  // de projeto atuais trazem o nome do projeto literal no caminho.
  const usaProjeto = regra.caminho_modelo.some((s) => s.includes("{PROJETO}"));
  if (usaProjeto && !(ctx.projeto ?? regra.projeto)) faltando.push("projeto");

  return faltando;
}

/** Descobre o escopo de uma pasta a partir do que gerou aquele segmento. */
function escopoDoSegmento(
  modeloOriginal: string,
  regra: RegraArquivamento,
  indice: number
): EscopoPasta {
  if (modeloOriginal.includes("{ANO}") || modeloOriginal.includes("{COMPETENCIA}") || modeloOriginal.includes("{COMPETENCIA_PASTA}")) {
    return "periodo";
  }
  if (modeloOriginal.includes("{PROJETO}")) return "projeto";

  // Regra de projeto: o segundo segmento do modelo é o nome do projeto,
  // escrito literalmente (ex.: CONFERENCIA_DE_CARTOES).
  if (regra.escopo === "projeto" && indice === 1) return "projeto";

  return regra.escopo === "interno" ? "estrutural" : "categoria";
}

/**
 * Expande a regra numa árvore de pastas concreta.
 *
 * Para regras com empresa, o primeiro nível é a pasta da EMPRESA e o
 * `caminho_modelo` começa logo abaixo dela. Para regras internas, o modelo
 * começa direto na raiz.
 */
export function expandirDestino(
  regra: RegraArquivamento,
  ctx: ContextoArquivamento
): DestinoResolvido | FalhaDestino {
  const faltando = conferirExigencias(regra, ctx);
  if (faltando.length > 0) {
    return {
      ok: false,
      erro: `A regra ${regra.codigo} exige: ${faltando.join(", ")}.`,
      faltando,
    };
  }

  if (ctx.competencia && !FORMATO_COMPETENCIA.test(ctx.competencia)) {
    return {
      ok: false,
      erro: `Competência "${ctx.competencia}" fora do formato AAAA-MM.`,
      faltando: ["competencia"],
    };
  }

  const ano = ctx.competencia ? ctx.competencia.slice(0, 4) : null;
  const projeto = normalizarSegmento(ctx.projeto ?? regra.projeto ?? "");

  const substituicoes: Record<string, string | null> = {
    ANO: ano,
    COMPETENCIA: ctx.competencia ?? null,
    COMPETENCIA_PASTA: ctx.competencia ? `${ctx.competencia.slice(5, 7)}.${ano}` : null,
    PROJETO: projeto || null,
  };

  const segmentos: SegmentoDestino[] = [];

  // A empresa entra na chave pelo UUID, não pelo nome: renomear a empresa
  // não pode quebrar o vínculo com a pasta já criada no Drive.
  const usaEmpresa = regra.exige_empresa;
  let prefixoChave = usaEmpresa ? `empresa:${ctx.empresaId}` : "estrutural";
  let caminhoAcumulado = "";

  if (usaEmpresa) {
    // 1º nível: o contêiner de clientes. A pasta de empresa NUNCA fica solta
    // na raiz — lá em cima só existem 00_INTERNO e os dois contêineres.
    const container = sanitizarNomeArquivo(ctx.pastaClientes ?? "").replace(/[/\\]/g, "_");
    if (!container) {
      return {
        ok: false,
        erro: "O contêiner de clientes ficou vazio após sanitização.",
        faltando: ["pastaClientes"],
      };
    }

    caminhoAcumulado = container;
    segmentos.push({
      nome: container,
      escopo: "estrutural",
      chaveLogica: `estrutural:${container}`,
      caminhoLogico: caminhoAcumulado,
    });

    // 2º nível: a empresa. A chave é SÓ `empresa:<uuid>`, sem o contêiner —
    // se o cliente vira inativo, o caminho desejado muda mas a chave não, e o
    // resolvedor reencontra a pasta existente em vez de criar uma segunda.
    // Mover a pasta é decisão humana, não efeito colateral de um upload.
    const nome = nomePastaEmpresa(ctx);
    if (!nome) {
      return {
        ok: false,
        erro: "Não foi possível montar o nome da pasta da empresa.",
        faltando: ["empresaCodigo"],
      };
    }

    caminhoAcumulado = `${caminhoAcumulado}/${nome}`;
    segmentos.push({
      nome,
      escopo: "empresa",
      chaveLogica: prefixoChave,
      caminhoLogico: caminhoAcumulado,
    });
  }

  for (const [indice, modelo] of regra.caminho_modelo.entries()) {
    let faltouPlaceholder: string | null = null;

    const expandido = modelo.replace(/\{([A-Z_]+)\}/g, (original, chave: string) => {
      if (!(chave in substituicoes)) return original;
      const valor = substituicoes[chave];
      if (valor === null) {
        faltouPlaceholder = chave;
        return original;
      }
      return valor;
    });

    if (faltouPlaceholder !== null) {
      const nomePlaceholder: string = faltouPlaceholder;
      return {
        ok: false,
        erro: `A regra ${regra.codigo} usa {${nomePlaceholder}} no caminho, mas o valor não foi informado.`,
        faltando: [nomePlaceholder.toLowerCase()],
      };
    }

    // Cada segmento é sanitizado isoladamente: nenhum valor consegue injetar
    // barra, `..` ou caractere de controle no caminho.
    const nome = sanitizarNomeArquivo(expandido).replace(/[/\\]/g, "_");
    if (!nome) {
      return {
        ok: false,
        erro: `O segmento "${modelo}" da regra ${regra.codigo} ficou vazio após sanitização.`,
        faltando: [],
      };
    }

    caminhoAcumulado = caminhoAcumulado ? `${caminhoAcumulado}/${nome}` : nome;
    prefixoChave = `${prefixoChave}:${nome}`;

    segmentos.push({
      nome,
      escopo: escopoDoSegmento(modelo, regra, indice),
      chaveLogica: prefixoChave,
      caminhoLogico: caminhoAcumulado,
    });
  }

  if (segmentos.length === 0) {
    return {
      ok: false,
      erro: `A regra ${regra.codigo} não produziu nenhuma pasta.`,
      faltando: [],
    };
  }

  const folha = segmentos[segmentos.length - 1];
  return {
    ok: true,
    segmentos,
    caminhoLogico: folha.caminhoLogico,
    chaveLogica: folha.chaveLogica,
  };
}

/**
 * Constrói segmentos para um caminho ESTRUTURAL literal, sem placeholder.
 *
 * Usado pela estrutura fixa (00_INTERNO, contêineres de clientes). Gera as
 * mesmas chaves que `expandirDestino` gera para regras internas, então a
 * pasta criada pelo bootstrap é exatamente a mesma que o arquivamento
 * reencontra depois — não são duas pastas com o mesmo nome.
 */
export function segmentosEstruturais(caminho: string[]): SegmentoDestino[] {
  const segmentos: SegmentoDestino[] = [];
  let chave = "estrutural";
  let acumulado = "";

  for (const bruto of caminho) {
    const nome = sanitizarNomeArquivo(bruto).replace(/[/\\]/g, "_");
    if (!nome) continue;

    chave = `${chave}:${nome}`;
    acumulado = acumulado ? `${acumulado}/${nome}` : nome;
    segmentos.push({ nome, escopo: "estrutural", chaveLogica: chave, caminhoLogico: acumulado });
  }

  return segmentos;
}

/**
 * Monta o nome final do arquivo a partir de `padrao_nome`.
 *
 * O nome ORIGINAL nunca é sobrescrito — ele continua em
 * documentos_operacionais.nome_original. Este é o nome do DESTINO.
 */
export function montarNomeArquivo(
  regra: RegraArquivamento,
  ctx: ContextoArquivamento
): { ok: true; nome: string } | FalhaDestino {
  const faltando = conferirExigencias(regra, ctx);
  if (!ctx.extensao) faltando.push("extensao");
  if (!ctx.tipoDocumento && regra.padrao_nome.includes("{TIPO_DOCUMENTO}")) {
    faltando.push("tipoDocumento");
  }
  if (faltando.length > 0) {
    return {
      ok: false,
      erro: `Para nomear pela regra ${regra.codigo} faltam: ${faltando.join(", ")}.`,
      faltando,
    };
  }

  if (ctx.dataDocumento && !FORMATO_DATA.test(ctx.dataDocumento)) {
    return {
      ok: false,
      erro: `Data do documento "${ctx.dataDocumento}" fora do formato AAAA-MM-DD.`,
      faltando: ["dataDocumento"],
    };
  }

  const empresaNome = ctx.empresaNome ?? ctx.empresaCodigo ?? "";

  // null = placeholder opcional sem valor: sai do nome em vez de virar lixo.
  const substituicoes: Record<string, string | null> = {
    CODIGO: normalizarSegmento(ctx.empresaCodigo ?? "") || null,
    EMPRESA: normalizarSegmento(empresaNome) || null,
    COMPETENCIA: ctx.competencia ?? null,
    TIPO_DOCUMENTO: normalizarSegmento(ctx.tipoDocumento ?? "") || null,
    INSTITUICAO: normalizarSegmento(ctx.instituicao ?? "") || null,
    PROJETO: normalizarSegmento(ctx.projeto ?? regra.projeto ?? "") || null,
    DATA_DOCUMENTO: ctx.dataDocumento ?? null,
    VERSAO: String(ctx.versao ?? 1),
    EXTENSAO: (ctx.extensao ?? "").toLowerCase().replace(/^\.+/, ""),
  };

  const bruto = regra.padrao_nome.replace(
    /\{([A-Z_]+)\}/g,
    (original, chave: string) => {
      if (!(chave in substituicoes)) return original;
      return substituicoes[chave] ?? "";
    }
  );

  // Colapsa o rastro dos placeholders removidos: "A__B" -> "A_B",
  // "A_.pdf" -> "A.pdf", "_A" -> "A".
  const limpo = bruto
    .replace(/_{2,}/g, "_")
    .replace(/_+\./g, ".")
    .replace(/^_+|_+$/g, "");

  const nome = sanitizarNomeArquivo(limpo);
  if (!nome) {
    return {
      ok: false,
      erro: `O nome gerado pela regra ${regra.codigo} ficou vazio.`,
      faltando: [],
    };
  }

  return { ok: true, nome };
}

File: Mini-Sistemas/arquivador_docs/arquivador/caminhos.py
"""
Onde o documento vai e como ele se chama.

Módulo PURO: não toca disco, não fala com rede. Recebe (regra + contexto) e
devolve os pedaços do caminho e o nome final. É a parte que decide, então é a
parte que precisa ser testável sem OneDrive nenhum.

Nenhum caminho é escrito aqui dentro. Tudo vem de `dados/regras.json`, que é
gerado a partir do banco.

PLACEHOLDERS DE CAMINHO: {ANO} {COMPETENCIA} {PROJETO}
PLACEHOLDERS DE NOME:    {CODIGO} {EMPRESA} {COMPETENCIA} {TIPO_DOCUMENTO}
                         {INSTITUICAO} {PROJETO} {DATA_DOCUMENTO} {VERSAO}
                         {EXTENSAO}

Placeholder opcional sem valor é REMOVIDO e os separadores que sobram são
colapsados — nunca sai "{INSTITUICAO}" literal nem "__" no meio do nome.
"""

from __future__ import annotations

import re
import unicodedata
from dataclasses import dataclass, field

FORMATO_COMPETENCIA = re.compile(r"^\d{4}-(0[1-9]|1[0-2])$")
FORMATO_DATA = re.compile(r"^\d{4}-(0[1-9]|1[0-2])-(0[1-9]|[12]\d|3[01])$")

# Nomes que o Windows recusa como arquivo ou pasta, mesmo com extensão.
RESERVADOS_WINDOWS = {
    "CON", "PRN", "AUX", "NUL",
    *(f"COM{i}" for i in range(1, 10)),
    *(f"LPT{i}" for i in range(1, 10)),
}


class ErroCaminho(Exception):
    """Falha previsível de montagem. Carrega o que faltou, para poder perguntar."""

    def __init__(self, mensagem: str, faltando: list[str] | None = None) -> None:
        super().__init__(mensagem)
        self.faltando = faltando or []


@dataclass(frozen=True)
class Regra:
    codigo: str
    nome: str
    escopo: str
    caminho_modelo: list[str]
    padrao_nome: str
    exige_empresa: bool = True
    exige_competencia: bool = False
    exige_instituicao: bool = False
    projeto: str | None = None
    subcategoria: str | None = None

    @staticmethod
    def de_dict(d: dict) -> "Regra":
        return Regra(
            codigo=d["codigo"],
            nome=d["nome"],
            escopo=d["escopo"],
            caminho_modelo=list(d["caminho_modelo"]),
            padrao_nome=d["padrao_nome"],
            exige_empresa=bool(d.get("exige_empresa", True)),
            exige_competencia=bool(d.get("exige_competencia", False)),
            exige_instituicao=bool(d.get("exige_instituicao", False)),
            projeto=d.get("projeto"),
            subcategoria=d.get("subcategoria"),
        )


@dataclass
class Contexto:
    """O que se sabe do documento. Campo ausente é None, nunca chute."""

    empresa_codigo: str | None = None
    empresa_nome: str | None = None
    pasta_clientes: str | None = None
    competencia: str | None = None
    instituicao: str | None = None
    tipo_documento: str | None = None
    data_documento: str | None = None
    projeto: str | None = None
    versao: int = 1
    extensao: str | None = None


@dataclass
class Destino:
    segmentos: list[str] = field(default_factory=list)

    @property
    def caminho_relativo(self) -> str:
        return "/".join(self.segmentos)


def normalizar(valor: str) -> str:
    """Sem acento, maiúsculo, só letra/número/_/-. Para nome de pasta e arquivo."""
    sem_acento = "".join(
        c for c in unicodedata.normalize("NFD", valor) if unicodedata.category(c) != "Mn"
    )
    limpo = re.sub(r"[^a-zA-Z0-9_-]+", "_", sem_acento)
    limpo = re.sub(r"_{2,}", "_", limpo).strip("_")
    return limpo.upper()


def sanitizar_segmento(valor: str) -> str:
    """
    Deixa um pedaço de caminho seguro para o Windows.

    Corta barra, caractere de controle e os símbolos que o Windows recusa; tira
    ponto e espaço do fim (o Explorer some com eles e o caminho deixa de bater);
    e desvia dos nomes reservados. É a barreira que impede um valor vindo do
    documento de escapar da pasta raiz.
    """
    base = re.split(r"[/\\]", valor)[-1]
    base = "".join(c for c in base if ord(c) >= 32 and c != "\x7f")
    base = re.sub(r'[<>:"|?*]', "_", base)
    base = base.strip().rstrip(". ")

    if base.upper().split(".")[0] in RESERVADOS_WINDOWS:
        base = f"_{base}"

    return base[:150]


def _competencia_valida(competencia: str) -> None:
    if not FORMATO_COMPETENCIA.match(competencia):
        raise ErroCaminho(
            f'Competência "{competencia}" fora do formato AAAA-MM.', ["competencia"]
        )


def conferir_exigencias(regra: Regra, ctx: Contexto) -> list[str]:
    """Devolve a LISTA do que falta — quem chama precisa saber o que perguntar."""
    faltando: list[str] = []

    if regra.exige_empresa:
        if not (ctx.empresa_codigo or ctx.empresa_nome):
            faltando.append("empresa_codigo")
        # Sem o contêiner a pasta do cliente cairia solta na raiz.
        if not ctx.pasta_clientes:
            faltando.append("pasta_clientes")

    if regra.exige_competencia and not ctx.competencia:
        faltando.append("competencia")
    if regra.exige_instituicao and not ctx.instituicao:
        faltando.append("instituicao")

    usa_projeto = any("{PROJETO}" in s for s in regra.caminho_modelo)
    if usa_projeto and not (ctx.projeto or regra.projeto):
        faltando.append("projeto")

    return faltando


def montar_destino(regra: Regra, ctx: Contexto) -> Destino:
    """Expande a regra na lista de pastas, da raiz até a pasta final."""
    faltando = conferir_exigencias(regra, ctx)
    if faltando:
        raise ErroCaminho(
            f"A regra {regra.codigo} exige: {', '.join(faltando)}.", faltando
        )

    if ctx.competencia:
        _competencia_valida(ctx.competencia)

    ano = ctx.competencia[:4] if ctx.competencia else None
    projeto = normalizar(ctx.projeto or regra.projeto or "") or None

    substituicoes = {
        "ANO": ano,
        "COMPETENCIA": ctx.competencia,
        "PROJETO": projeto,
    }

    segmentos: list[str] = []

    if regra.exige_empresa:
        # 1º nível: contêiner de clientes. 2º: a empresa. A pasta do cliente
        # nunca fica solta na raiz.
        container = sanitizar_segmento(ctx.pasta_clientes or "")
        if not container:
            raise ErroCaminho("Contêiner de clientes vazio.", ["pasta_clientes"])
        segmentos.append(container)

        empresa = normalizar(ctx.empresa_codigo or ctx.empresa_nome or "")
        if not empresa:
            raise ErroCaminho("Nome da pasta da empresa vazio.", ["empresa_codigo"])
        segmentos.append(empresa)

    for modelo in regra.caminho_modelo:
        faltou: str | None = None

        def troca(m: re.Match[str]) -> str:
            nonlocal faltou
            chave = m.group(1)
            if chave not in substituicoes:
                return m.group(0)
            valor = substituicoes[chave]
            if valor is None:
                faltou = chave
                return m.group(0)
            return valor

        expandido = re.sub(r"\{([A-Z_]+)\}", troca, modelo)

        if faltou:
            raise ErroCaminho(
                f"A regra {regra.codigo} usa {{{faltou}}} no caminho, "
                "mas o valor não foi informado.",
                [faltou.lower()],
            )

        seguro = sanitizar_segmento(expandido)
        if not seguro:
            raise ErroCaminho(
                f'O segmento "{modelo}" da regra {regra.codigo} ficou vazio.'
            )
        segmentos.append(seguro)

    if not segmentos:
        raise ErroCaminho(f"A regra {regra.codigo} não produziu nenhuma pasta.")

    return Destino(segmentos=segmentos)


def montar_nome(regra: Regra, ctx: Contexto) -> str:
    """Nome final do arquivo. O nome original NUNCA é sobrescrito no registro."""
    faltando = conferir_exigencias(regra, ctx)
    if not ctx.extensao:
        faltando.append("extensao")
    if "{TIPO_DOCUMENTO}" in regra.padrao_nome and not ctx.tipo_documento:
        faltando.append("tipo_documento")
    if faltando:
        raise ErroCaminho(
            f"Para nomear pela regra {regra.codigo} faltam: {', '.join(faltando)}.",
            faltando,
        )

    if ctx.data_documento and not FORMATO_DATA.match(ctx.data_documento):
        raise ErroCaminho(
            f'Data "{ctx.data_documento}" fora do formato AAAA-MM-DD.',
            ["data_documento"],
        )

    empresa_nome = ctx.empresa_nome or ctx.empresa_codigo or ""
    extensao = (ctx.extensao or "").lower().lstrip(".")

    # None = placeholder opcional sem valor: sai do nome em vez de virar lixo.
    substituicoes: dict[str, str | None] = {
        "CODIGO": normalizar(ctx.empresa_codigo or "") or None,
        "EMPRESA": normalizar(empresa_nome) or None,
        "COMPETENCIA": ctx.competencia,
        "TIPO_DOCUMENTO": normalizar(ctx.tipo_documento or "") or None,
        "INSTITUICAO": normalizar(ctx.instituicao or "") or None,
        "PROJETO": normalizar(ctx.projeto or regra.projeto or "") or None,
        "DATA_DOCUMENTO": ctx.data_documento,
        "VERSAO": str(ctx.versao),
        "EXTENSAO": extensao,
    }

    bruto = re.sub(
        r"\{([A-Z_]+)\}",
        lambda m: substituicoes.get(m.group(1), m.group(0)) or "",
        regra.padrao_nome,
    )

    # Colapsa o rastro do que foi removido: "A__B" -> "A_B", "A_.pdf" -> "A.pdf".
    limpo = re.sub(r"_{2,}", "_", bruto)
    limpo = re.sub(r"_+\.", ".", limpo).strip("_")

    nome = sanitizar_segmento(limpo)
    if not nome:
        raise ErroCaminho(f"O nome gerado pela regra {regra.codigo} ficou vazio.")

    return nome


def trocar_versao_no_nome(nome: str, versao: int) -> str:
    """Troca o _vN antes da extensão. Se não houver, acrescenta."""
    novo, trocas = re.subn(r"_v\d+(\.[^.]+)$", rf"_v{versao}\1", nome)
    if trocas:
        return novo

    if "." in nome:
        base, _, ext = nome.rpartition(".")
        return f"{base}_v{versao}.{ext}"
    return f"{nome}_v{versao}"
